openpype/pipeline/anatomy.py

```diff
@@ -606,6 +607,32 @@ def normalized(self):
         return self.__class__(tmp, self.rootless)
 
 
+class AnatomyStringTemplate(StringTemplate):
+    """String template which has access to anatomy."""
+
+    def __init__(self, anatomy_templates, template):
+        self.anatomy_templates = anatomy_templates
+        super(AnatomyStringTemplate, self).__init__(template)
+
+    def format(self, data):
+        """Format template and add 'root' key to data if not available.
+
+        Args:
+            data (dict[str, Any]): Formatting data for template.
+
+        Returns:
+            AnatomyTemplateResult: Formatting result.
+        """
+
+        anatomy_templates = self.anatomy_templates
+        if not data.get("root"):
+            data = copy.deepcopy(data)
+            data["root"] = anatomy_templates.anatomy.roots
```

Comment:
I suspect that this `format` per string attribute is also called when doing a `anatomy.format()` call? Which makes it look like now we're deep copying the input data per string attribute if the data didn't include the `root` key which will be much slower since we have quite a few string attibutes. Right?

I'm also looking at the code that changed here but it didn't seem to do this "automatic ingesting of root" before either - it just required it. Should we keep it that explicit?

Reply:
> Which makes it look like now we're deep copying the input data per string attribute if the data didn't include the root key which will be much slower since we have quite a few string attibutes. Right?

It does copy of data only if `"root"` is not in source data, when you call `anatomy.format(...)` it will fill the `"root"`. It is not explicit because of that.

Could be made explicit and faster by changing to this:
```python
if data.get("root") != anatomy_templates.anatomy.roots:
    data = copy.deepcopy(data)
    data["root"] = anatomy_templates.anatomy.roots
```

Reply:
If `anatomy.format()` internally already pre-sets the `root` key then I'm fine with the current logic. I was just worried that now suddenly it'd start setting it for each string it's formatting.